### ISSUE TITLE
Javascript crash 

### ISSUE THREAD

**Issue by ajankov** [annulen/webkit#787]
Many websites still crash in Windows 10, Visual Studio build of qtwebkit . You can find a short list of web sites below. The crashes are in general in JavaScript execution. Some other websites never finish loading. The JavaScript issue seems to have the same root cause already reported at https://github.com/annulen/webkit/issues/741 and the backtrace is very similar. 
If you can find some time to investigate this and suggest a patch to fix the crashes that would be very much appreciated. 

Crashing web sites :

https://issuu.com/adaniieell/docs/arduino_y_matlab
https://noticiasdechiapas.com.mx/noticia.php?item=932
https://mundocontact.com/los-15-mejores-gadgets-del-ano/
https://revistagloborural.globo.com/vida-na-fazenda/noticia/2015/05/como-fazer-o-carneiro-hidraulico.html
https://www.bls.gov/ooh/a-z-index.htm#A
https://www.quikr.com/jobs/office-boy-helper-ghaziabad-W0QQAdIdZ311750648
https://www.pure-design.de/
https://www.yoamoelgolf.mx/blog/bbva-bancomer-patrocinar%C3%A1-el-torneo-interclubes-de-golf-2018
https://www.auto-swiat.pl/testy/testy-nowych-samochodow/hyundai-i30-16-crdi-fastback-cos-za-cos-or-test/95kewq9

Never finish loading web sites:

https://canvasjs.com/javascript-charts/multi-series-chart/
http://mama.mwave.me/kr/main

**Comment by ajankov** [comment]
The issues are more frequent for spanish websites. https://www.elmundo.es/espana.html or https://www.marca.com/ never finish loading and are hanging.

**Comment by ajankov** [comment]
Any suggestion to patch these ?

**Comment by annulen** [comment]
Are you using debug build?

**Comment by ajankov** [comment]
The JavaScript crashes are reproducible in debug and release. I think they are also related to bug report https://github.com/annulen/webkit/issues/741 . A common thing for many websites in the list above is they are created with https://issuu.com/ publishing platform. I have tried a lot to find a fix or at least to handle crashes but I did not succeed. Your help to solve this problem would be highly appreciated. Currently I cannot continue to use qtwebkit because of this problem.

**Comment by ajankov** [comment]
Do these JavaScript problems reported here and in https://github.com/annulen/webkit/issues/741  reproduce in your builds? 
Will you consider looking into them? I think you are the only one who could solve such problems, they seem to be pretty advanced and they are above my expertise.
Do you think they are caused by upstream WebKit? Maybe you could point to a possible patch available in upstream WebKit? Thanks.

**Comment by annulen** [comment]
I don't use windows so I have no idea

**Comment by ajankov** [comment]
Thanks for your answer. Should I understand that you don't have any intention to offer any help to solve this? Maybe with your expertise it is something simple and you can figure out quickly a possible solution to avoid these crashes? Otherwise I don't think the current build can be used on Windows unfortunately. Thanks.

**Comment by annulen** [comment]
Please try thingds suggested in https://github.com/annulen/webkit/wiki/Debugging-issues-with-JavaScript-execution

**Comment by ajankov** [comment]
Is it possible to disable JIT on WIndows?

**Comment by annulen** [comment]
Article linked above describes how to do that

**Comment by ajankov** [comment]
Some time ago you said that 'Building without JIT does not work on Windows'. Does this mean that you have to build with JIT and later disable it? Could you tell me what is the supported procedure on Windows to disable JIT?

**Comment by annulen** [comment]
>Does this mean that you have to build with JIT and later disable it?

Yes. See https://github.com/annulen/webkit/wiki/Debugging-issues-with-JavaScript-execution

**Comment by ajankov** [comment]
I tried to alternatively disable JIT and LLINT and it doesn't change anything.

**Comment by annulen** [comment]
@herbmillerjr do these crashes reproduce with your build?

**Comment by ajankov** [comment]
Thanks!

**Comment by ajankov** [comment]
I tested with the latest binary release of Otter browser for Windows I downloaded from https://otter-browser.org/  and it has the same crashes on the same websites above. You can check this if you have a Windows 10 machine.

**Comment by annulen** [comment]
Please try MiniBrowser.exe from https://s3-us-west-2.amazonaws.com/archives.webkit.org/wincairo-x86_64-release/243693.zip (you'll also need to unpack https://github.com/WebKitForWindows/WinCairoRequirements/releases/download/v2019.03.28/WinCairoRequirements.zip and make sure all DLLs are in one dir with MiniBrowser.exe)

**Comment by ajankov** [comment]
I've just tried and in general the mini browser you asked me to test works well with majority of the websites I have tested. 

I saw an issue only with the first web site from the list. Many elements like images of that site were not loaded  and when I created a LegacyWebKit windows from file menu then in that case the browser crashes. For all the other websites I tested it works both in default windows of the mini browser and in the legacy window.

**Comment by annulen** [comment]
Thanks for you feedback

**Comment by ajankov** [comment]
Does the issue seem to be caused by the upstream WebKit version used in qtwebkit? If there is anything else you would like me to check please let me know. 

Do you think there is a real chance to update the qtwebkit with a newer upstream webkit or the effort is too big for you if nobody else from upstream WebKit is involved in QtWebKit port? I remember some time ago you you said you are preparing a new build machine for a new version. Did you change the plan in meantime?

I really understand that is a huge task to update the webkit and sharing a few thoughts you have at this moment about the future would be highly appreciated by me and by many other people still interested in QtWebKit port. 

Thank you.

**Comment by annulen** [comment]
>Does the issue seem to be caused by the upstream WebKit version used in qtwebkit?

Either this, or different build environment (compiler version/build options/runtime/etc). That's why I've asked @herbmillerjr to check it as well

BTW, are you using VS2017 or 2015? 32-bit or 64-bit target? Did these crashes reproduce with builds from qt.io, e.g. http://download.qt.io/snapshots/ci/qtwebkit/5.212/latest/qtwebkit/ (which uses Qt 5.10)

>

Do you think there is a real chance to update the qtwebkit with a newer upstream webkit or the effort is too big for you if nobody else from upstream WebKit is involved in QtWebKit port? I remember some time ago you you said you are preparing a new build machine for a new version. Did you change the plan in meantime?

Plan is still here, however I'm not sure that Windows will be supported. To make this happen, either someone needs to step up as a volunteer maintainer of Windows platform, or there should be steady stream of donations which would allow hiring part-time developer to do maintenance work. That's not to say that I'm going to stay away from any Windows-specific issues, if such maintainer is found, but I'd like to reduce my interference with this platform as much as its possible. Otherwise, it can only be supported at level of "If it compiles, it is good; if it boots up, it is perfect", which is how it works now. It's misleading for users, and it still requires effort to make things buildable

**Comment by ajankov** [comment]
Thanks for sharing this. 

The issue reproduces with Qt 5.7.1 and later with VS2015 and VS2017.

For the snapshots

**Comment by annulen** [comment]
>I remember some time ago you you said you are preparing a new build machine for a new version. 

FYI, I've prepared that machine, and environment for running webkit tests is almost ready (there are a few remaining issues)

**Comment by ajankov** [comment]
Do you mean you are preparing for testing an updated version of qtwebkit?

**Comment by annulen** [comment]
>I saw an issue only with the first web site from the list. Many elements like images of that site were not loaded

I guess that's expected - WinCairo folks warned me that there are rendering issues with current trunk

**Comment by annulen** [comment]
Not so fast. To test updated version, I need to make old version pass tests on new system configuration at first

**Comment by ajankov** [comment]
For the snapshots from qt.io are there any prebuilt test browsers like qttestbrowser or minibrowser I can use for testing?

**Comment by annulen** [comment]
Unfortunately no, you'll have to compile something yourself

**Comment by annulen** [comment]
BTW, could you remind your IRC nick please?

**Comment by ajankov** [comment]
I did not use IRC before.

**Comment by ajankov** [comment]
Have you been able to reproduce the crashes in a Windows build? It reproduces for me with VS 2015, 2017 with Qt 5.7.1 and later. Since the Windows builds are not actively tested and supported I think it does not make any sense to further use the library on Windows.

**Comment by annulen** [comment]
This issue won't be fixed until WebKit upgrade is finished, and it's unclear if there will be any Windows support after that point

**Comment by ajankov** [comment]
Should I understand that the issue was reproduced?

**Comment by annulen** [comment]
Yes, I can reproduce it. Various slow paths crash when called. I have no guesses what can cause this, beyound maybe stack alignment issues, however I see no windows-specific patches in trunk addressing such problem so there is nothing to cherry-pick

**Comment by annulen** [comment]
What's especially weird is that it reproduces on 64-bit build, where possibility of ABI mismatch issues is lowest, as there are no bizarre calling conventions involved (i.e. fastcall)

**Comment by annulen** [comment]
Actually, stack seems to be aligned properly

**Comment by annulen** [comment]
Also, crashes are not 100% reproducible, sometimes page works fine, another time crashes

**Comment by ajankov** [comment]
Thank you for taking the time to investigate this. At least now I know that I don't do something wrong. 
The crashes are consistently reproducible for me.

Does it work without crashes for you on Linux? If the Linux build works well it means the issue is Windows specific and since the latest version of WinCairo works well I would conclude that there was a fix in the meantime, maybe not Windows specific but something that affected all platforms and also corrected the issue on Windows. I think you know much better than me how to interpret these scenarios.

**Comment by annulen** [comment]
>Does it work without crashes for you on Linux?

Yes. Otherwise there would be a queue of angry users in bug tracker ;)

>If the Linux build works well it means the issue is Windows specific and since the latest version of WinCairo works well I would conclude that there was a fix in the meantime, maybe not Windows specific but something that affected all platforms and also corrected the issue on Windows

There were lots of groundbreaking changes in JS engine, so it's definitely a possibility

**Comment by ajankov** [comment]
I understand. If something that might fix this problem will ever cross your mind and you don't have enough time for testing the idea just let me know and I'll be glad to try it.

**Comment by ajankov** [comment]
Since this might be the last supported release for Windows could you look once more into this issue and maybe you can find a way to avoid these crashes even if the final rendering would not be correct, like catching an exception in JavaScript VM or something else? Thanks.

**Comment by annulen** [comment]
"Supported" is an exaggeration. It has a word "alpha" in a version for a reason. Currently windows is "supported" in a sense that QtWebKit compiles there, and it's quite likely that the next version will also compile on Windows, at least with a few trivial patches.

**Comment by ajankov** [comment]
Could you try to suggest a way to avoid or at least to ignore those crashes please?

**Comment by ajankov** [comment]
The final crash is in function below on the bold line from LowLevelInterpreter64.asm for a 64-bit build and in the same function on the bold line for a 32-bit build

macro cCall2(function)
    checkStackPointerAlignment(t4, 0xbad0c002)
    if X86_64 or ARM64
        call function
    elsif X86_64_WIN
        # Note: this implementation is only correct if the return type size is > 8 bytes.
        # See macro cCall2Void for an implementation when the return type <= 8 bytes.
        # On Win64, when the return type is larger than 8 bytes, we need to allocate space on the stack for the return value.
        # On entry rcx (a0), should contain a pointer to this stack space. The other parameters are shifted to the right,
        # rdx (a1) should contain the first argument, and r8 (a2) should contain the second argument.
        # On return, rax contains a pointer to this stack value, and we then need to copy the 16 byte return value into rax (r0) and rdx (r1)
        # since the return value is expected to be split between the two.
        # See http://msdn.microsoft.com/en-us/library/7572ztz4.aspx
        move a1, a2
        move a0, a1
        subp 48, sp
        move sp, a0
        addp 32, a0
        **call function**
        addp 48, sp
        move 8[r0], r1
        move [r0], r0
    elsif C_LOOP
        cloopCallSlowPath function, a0, a1
    else
        error
    end
end

macro cCall2(function)
    if ARM or ARMv7 or ARMv7_TRADITIONAL or MIPS or SH4
        call function
    elsif X86 or X86_WIN
        **subp 8, sp**
        push a1
        push a0
        call function
        addp 16, sp
    elsif C_LOOP
        cloopCallSlowPath function, a0, a1
    else
        error
    end
end

**Comment by annulen** [comment]
Sorry, I'm not really competent to fix this issue. If I start working on it, it will take time to learn rules of windows ABIs and then have lengthy debug sessions trying to find out what this damn thing doesn't like in call site and which kind of memory corruption has lead to it. Yout options are: 1) do it yourself; 2) hire someone to do this for you, e.g. you can try to contact @vitallium; 3) help me to reach goal #2 on https://www.patreon.com/annulen so that I can hire someone to do this, or roll new QtWebKit version faster which may be free from this issue

**Comment by Vitallium** [comment]
Hi! I will have a look into this problem. I will continue supporting QtWebkit for Windows. I didn't have time for it in the past due to my family stuff. But now I am back.

**Comment by ajankov** [comment]
Thanks for looking into this !

**Comment by ajankov** [comment]
Hi Vitaly, have you had the chance to take a look at this ? Can I do anything to help you? Thanks.

**Comment by ajankov** [comment]
Maybe someone involved in WebKit upstream development can help us with a suggestion since it seems to be such a difficult issue? Maybe it was a known issue with the WebKit ...

**Comment by vitallium** [comment]
Hi, sorry. I am still traveling. Didn't have time to look into this issue.

**Comment by annulen** [comment]
@ajankov Update on situation with QtWebKit:
* New branch will certainly have some suppport of Windows. It will compile there, and very basic functionality of browser will work without critical bugs (no guarantees about advanced features like multimedia, hardware acceleration, WebGL, QML, etc.). If I'll have time to mess with Qt CI, there may be even precompiled binaries built with Qt 5.13 or 5.12
* You'll get early preview version in approximately 2 months from now. It's hard to guess if it's quality will be enough for your project, but you'll be able to start testing and report bugs
* If you are willing to speed up development, or need advanced features on Windows, or need priority fixing of bugs[*], you should consider funding of QtWebKit, e.g. via Pateron page (https://www.patreon.com/annulen), other funding platforms may be considered as wel (e.g. GitHub is launching its own program)

[*] Tiny font disclaimer: Windows-specific bugs will be fixed after those affecting Linux. If there any bugs caused by upstream code and reproduce in then-current WebKit trunk, such bugs are not guaranteed to be fixed by us, but they will be properly reported to upstream.

**Comment by ajankov** [comment]
Any chance you'll look into this?

**Comment by ajankov** [comment]
I know you are preparing an update but alpha 2 was quite good except this crash. The new build might come with new issues until it gets stable and the Windows support is quite uncertain.

**Comment by annulen** [comment]
>I know you are preparing an update but alpha 2 was quite good except this crash.

Not counting myriads of potential security issues.

I think at this point working on update would be much more productive use of time then working on this issue.

>Windows support is quite uncertain

One of sponsors is interested in Windows support, so it certainly will be implemented. If you are willing to speed things up, you are welcome to join development team, or become our sponsor.

**Comment by ajankov** [comment]
Do you have an updated estimation for a release or preview build for Windows based on the new upstream qtwebkit? I saw there was a lot of activity around it and I was wondering if it's still feasible. Thanks.

**Comment by annulen** [comment]
Current priority is to release Linux version first (though issues which are currently in work are platform-independent). If you want to accelerate development, you are invited to support the project at https://www.patreon.com/annulen.

**Comment by dugo** [comment]
Hi,

sorry for the delay.

I have been spent these days compiling and testing the branch qtwebkit-dev-wip with Qt 5.12.2

As I have checked, the branch fixes some bugs and strange behaviours. Good work!

But as you mentioned, it still includes some bugs that produce hangings and other issues, and it isn't usable for production.

Is possible to fix these problems? If yes, can it be fixed in a mid/long term?

Thank you so much

**Comment by annulen** [comment]
>it isn't usable for production
That's correct

>Is possible to fix these problems?
Definitely
